File: Algorithms/Sorting/MergeSort/MergeSort.py

def mergeSort(arr):
    if len(arr) > 1:

        middle = len(arr) // 2
        leftArr = arr[:middle]
        rightArr = arr[middle:]

        mergeSort(leftArr)
        mergeSort(rightArr)

        merge(arr, leftArr, rightArr)

def merge(arr,leftArr, rightArr):
    index = leftIndex = rightIndex = 0
    while leftIndex < len(leftArr) and rightIndex < len(rightArr):
        if(leftArr[leftIndex] <= rightArr[rightIndex]):
            arr[index] = leftArr[leftIndex]
            leftIndex += 1
        else:
            arr[index] = rightArr[rightIndex]
            rightIndex += 1
        index += 1

    #fill remaining items
    while leftIndex < len(leftArr):
        arr[index] = leftArr[leftIndex]
        leftIndex += 1
        index += 1

    while rightIndex < len(rightArr):
        arr[index] = rightArr[rightIndex]
        rightIndex += 1
        index += 1

File: Algorithms/Sorting/MergeSort/MergeSort.js
class MergeSort {
  sort(arr) {
    if (arr.length > 1) {
      const middle = Math.floor(arr.length / 2);
      const leftArr = arr.slice(0, middle);
      const rightArr = arr.slice(middle, arr.length);

      this.sort(leftArr);
      this.sort(rightArr);

      this.merge(arr, leftArr, rightArr);
    }
  }

  merge(arr, leftArr, rightArr) {
    let index = 0,
      leftIndex = 0,
      rightIndex = 0;
    while (leftIndex < leftArr.length && rightIndex < rightArr.length) {
      if (leftArr[leftIndex] <= rightArr[rightIndex])
        arr[index] = leftArr[leftIndex++];
      else arr[index] = rightArr[rightIndex++];
      index++;
    }

    while (leftIndex < leftArr.length) arr[index++] = leftArr[leftIndex++];

    while (rightIndex < rightArr.length) arr[index++] = rightArr[rightIndex++];
  }
}

module.exports = MergeSort;
